src/main/java/de/tum/in/www1/artemis/service/competency/CompetencyProgressService.java

```diff
@@ -217,26 +204,187 @@ public CompetencyProgress updateCompetencyProgress(Long competencyId, User user)
+     * If the proportion of achieved points in the given difficulty is higher than the proportion of points in the competency, the confidence should be higher.
+     *
+     * @param participantScores  the participant scores for the exercises linked to the competency
+     * @param exerciseInfos      the information about the exercises linked to the competency
+     * @param achievedPoints     the total points achieved by the user in the competency
+     * @param pointsInCompetency the total points in the competency
+     * @param difficultyLevel    the difficulty level to calculate the confidence for
+     * @return the difficulty confidence heuristic for the given difficulty
+     */
+    private double calculateDifficultyConfidenceHeuristicForDifficulty(@NotNull Set<CompetencyExerciseMasteryCalculationDTO> participantScores,
+            @NotNull Set<CompetencyExerciseMasteryCalculationDTO> exerciseInfos, double achievedPoints, double pointsInCompetency, DifficultyLevel difficultyLevel) {
+
+        double achievedPointsInDifficulty = participantScores.stream().filter(info -> info.difficulty() == difficultyLevel)
+                .mapToDouble(CompetencyExerciseMasteryCalculationDTO::lastPoints).sum();
+        double pointsInCompetencyInDifficulty = exerciseInfos.stream().filter(info -> info.difficulty() == difficultyLevel)
+                .mapToDouble(CompetencyExerciseMasteryCalculationDTO::maxPoints).sum();
+
+        double proportionOfAchievedPointsInDifficulty = achievedPointsInDifficulty / achievedPoints;
+        double proportionOfPointsInCompetencyInDifficulty = pointsInCompetencyInDifficulty / pointsInCompetency;
+
+        return proportionOfAchievedPointsInDifficulty - proportionOfPointsInCompetencyInDifficulty;
+    }
+
+    /**
+     * Calculate the quick solve confidence heuristic for the given user in a competency based on the exercises linked to the competency.
+     * If the user has solved a high proportion of programming exercises with a high score in a short amount of time, the confidence should be higher.
+     *
+     * @param participantScores the participant scores for the exercises linked to the competency
+     * @return The quick solve confidence heuristic
+     */
+    private double calculateQuickSolveConfidenceHeuristic(@NotNull Set<CompetencyExerciseMasteryCalculationDTO> participantScores) {
+        Set<CompetencyExerciseMasteryCalculationDTO> programmingParticipationScores = participantScores.stream()
+                .filter(CompetencyExerciseMasteryCalculationDTO::isProgrammingExercise).collect(Collectors.toSet());
+
+        if (programmingParticipationScores.isEmpty()) {
+            return 0;
+        }
+
+        double numberOfQuicklySolvedProgrammingExercises = programmingParticipationScores.stream()
+                .filter(info -> info.lastScore() >= MIN_SCORE_GREEN && info.submissionCount() <= 3).count();
```

Comment:
Is the 3 related to the previous check in `calculateRecencyConfidenceHeuristic`? If yes I think they should have the same variable, otherwise I would be in favor of different variables to make clear that these values are independent